src/jit/hwintrinsicxarch.cpp

```diff
@@ -295,6 +304,130 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
     return arg;
 }
 
+//------------------------------------------------------------------------
+// immUpperBoundOfHWIntrinsic: get the max imm-value of non-full-range IMM intrinsic
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//
+// Return Value:
+//     the max imm-value of non-full-range IMM intrinsic
+//
+int Compiler::immUpperBoundOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(categoryOfHWIntrinsic(intrinsic) == HW_Category_IMM);
+    HWIntrinsicFlag flags = flagsOfHWIntrinsic(intrinsic);
+    if ((flags & HW_Flag_FullRangeIMM) != 0)
+    {
+        return 255;
+    }
+
+    switch (intrinsic)
+    {
+        case NI_AVX_Compare:
+        case NI_AVX_CompareScalar:
+            return 31; // enum FloatComparisonMode has 32 values
+
+        default:
+            assert(!"Not defined non-full-range inn-intrinsics");
+            return 255;
+    }
+}
+
+//------------------------------------------------------------------------
+// impNonConstFallback: convert certain SSE2/AVX2 shift intrinsic to its semantic alternative when the imm-arg is
```

Comment:
This seems ok/like a good direction to me. But want to make sure that this is ok/inline with the overall feature design, @CarolEidt?

Reply:
It seems reasonable to me, but there needs to be more clarity with respect to `HW_Category_IMM`, `HW_Flag_MaybeIMM` and `HW_Flag_NoJmpTableIMM`. In particular, the comments say that `HW_Category_IMM` intrinsics *require* immediate value, but there are entries in the table that have that set, as well as `HW_Flag_MaybeIMM`.